File: ruby/binary-search/binary_search.rb
class BinarySearch
  attr_reader :list

  def initialize(list)
    raise ArgumentError unless list == list.sort

    @list = list
  end

  def middle(left = 0, right = @list.length - 1)
    (left + right) / 2
  end

  def search_for(elem, left = 0, right = @list.length - 1)
    return nil if left > right

    mid = middle(left, right)

    return mid if elem == @list[mid]

    right = mid - 1 if elem < @list[mid]
    left  = mid + 1 if elem > @list[mid]
    search_for elem, left, right
  end
end

File: kotlin/binary-search/src/main/kotlin/BinarySearch.kt
object BinarySearch {
    tailrec fun search(list: List<Int>, item: Int): Int {
        if (list.isEmpty())
            throw NoSuchElementException()

        var left = 0
        var right = list.size - 1
        val mid = right / 2

        when {
            item == list[mid] -> return mid
            item  < list[mid] -> right = mid - 1
            else              -> left = mid + 1
        }

        return left + search(list.subList(left, right + 1), item)
    }
}
